apps/web/src/components/Paymaster/GasTokenModal.tsx

```diff
@@ -0,0 +1,375 @@
+import memoize from 'lodash/memoize'
+import { useCallback, useEffect, useMemo } from 'react'
+import { FixedSizeList } from 'react-window'
+import { useNativeBalances, useTokenBalancesWithLoadingIndicator } from 'state/wallet/hooks'
+import styled from 'styled-components'
+import { Address } from 'viem'
+import { useAccount, useConfig } from 'wagmi'
+
+import { SmartRouterTrade } from '@pancakeswap/smart-router'
+import { Currency, TradeType } from '@pancakeswap/swap-sdk-core'
+import { watchAccount } from '@wagmi/core'
+import { DEFAULT_PAYMASTER_TOKEN, paymasterInfo, paymasterTokens } from 'config/paymaster'
+import { useGasToken } from 'hooks/useGasToken'
+import { RouteDisplayEssentials } from 'views/Swap/V3Swap/components'
+import { TradeEssentialForPriceBreakdown } from 'views/Swap/V3Swap/utils/exchange'
+
+// Selector Styles
+const GasTokenSelectButton = styled(Button).attrs({ variant: 'text', scale: 'xs' })`
+  padding: 18px 0 18px 6px;
+`
+
+// Modal Styles
+const StyledModalContainer = styled(ModalContainer)`
+  width: 100%;
+  min-height: calc(var(--vh, 1vh) * 60);
+  ${({ theme }) => theme.mediaQueries.md} {
+    min-height: auto;
+    min-width: 320px;
+    max-width: 420px !important;
+  }
+`
+
+const StyledModalBody = styled(ModalBody)`
+  overflow-y: auto;
+  -ms-overflow-style: none;
+  scrollbar-width: none;
+  &::-webkit-scrollbar {
+    display: none;
+  }
+`
+
+const StyledFooterText = styled(Text)`
+  padding: 10px 0 20px 0;
+`
+
+const FixedHeightRow = styled.div<{ $disabled: boolean }>`
+  height: 56px;
+  display: flex;
+  align-items: center;
+  padding: 0 16px;
+
+  cursor: ${({ $disabled }) => !$disabled && 'pointer'};
+
+  &:hover {
+    background-color: ${({ theme, $disabled }) => !$disabled && theme.colors.background};
+  }
+
+  ${({ $disabled }) =>
+    $disabled &&
+    `
+    opacity: 0.5;
+    user-select: none;
+`}
+`
+
+const StyledBalanceText = styled(Text).attrs({ small: true })`
+  white-space: nowrap;
+  overflow: hidden;
+  text-overflow: ellipsis;
+`
+
+const Badge = styled.span`
+  font-size: 14px;
+  padding: 1px 6px;
+  border-radius: ${({ theme }) => theme.radii['32px']};
+  color: ${({ theme }) => theme.colors.invertedContrast};
+  background-color: ${({ theme }) => theme.colors.success};
+`
+
+const SameTokenWarningBox = styled(Box)`
+  font-size: 13px;
+  background-color: #ffb2371a;
+  padding: 10px;
+  margin: 5px 0 8px;
+  color: ${({ theme }) => theme.colors.yellow};
+  border: 1px solid ${({ theme }) => theme.colors.yellow};
+  border-radius: ${({ theme }) => theme.radii['12px']};
+`
+
+const StyledWarningIcon = styled(WarningIcon)`
+  fill: ${({ theme }) => theme.colors.yellow};
+`
+
+type Trade = TradeEssentialForPriceBreakdown &
+  Pick<SmartRouterTrade<TradeType>, 'tradeType'> & {
+    routes: RouteDisplayEssentials[]
+  }
+
+interface GasTokenModalProps {
+  trade?: Trade | null
+}
+
+export const GasTokenModal = ({ trade }: GasTokenModalProps) => {
```

Comment:
Shall we rename? This component is not only about modal but also the selector with label

Reply:
`GasTokenSelector`?
Both button and modal are part of selection, so will cover both naming cases